### PR TITLE
feat: add dataplane metrics for control plane

### PR DESCRIPTION
**Pull request by omegabytes**
Adds the following metrics:
    * duration (s) to establish initial connection between dp and cp
    * duration (s) of individual config mutations
    * duration (s) of total aggregate config mutations
    * duration (s) of version compat cache miss path

### CODE REVIEW

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [inline internal/server/kong/ws/config/configuration.go:81]: This (and other instances of this) can be simplified: 😃 
```suggestion
		mutationStartTime := time.Now()
		err := m.Mutate(ctx, MutatorOpts{ClusterID: clusterID},
			configTable)
		metrics.Histogram(
			"data_plane_config_mutation_time_individual",
			float64(time.Since(mutationStartTime).Milliseconds()),
			metrics.Tag{Key: "status", Value: lo.Ternary(err == nil, "success", "fail")},
			metrics.Tag{Key: "mutator_name", Value: m.Name()},
		)
		if err != nil {
			return Content{}, err
		}
```

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [inline internal/server/kong/ws/config/configuration.go:60]: Can we follow standard Prometheus conventions: 

1. Naming this metric series: `data_plane_config_mutation_individual_duration_seconds`
    - We're not emitting the entire mutation process as a metric. Do we want to be doing that too?
2. Altering to push seconds instead (with a decimals)

File: internal/server/kong/ws/server.go
Comment (tjasko) [inline internal/server/kong/ws/server.go:126]: Related to the other comments, can we use seconds here & rename to `data_plane_connect_duration_seconds`.

Also, should we rename this to `data_plane_websocket_connect_duration_seconds` to be a bit more specific?

File: internal/server/kong/ws/server.go
Comment (tjasko) [inline internal/server/kong/ws/server.go:90]: Should we move this right above `m, err := h.authenticator.Authenticate(r)`? As the way it currently is, it'd capture a bit more than just connecting.

File: internal/server/kong/ws/node.go
Comment (tjasko) [inline internal/server/kong/ws/node.go:191]: This won't do what you expect, as it's grabbing the size of the entire `config.Content` object & not the size of the actual config.

Let's be sure we're capturing the size of the JSON payload being sent to the data plane, and ideally, it's uncompressed size as that's what matters. Let's modify the `config.Content` struct to include the size in bytes as well, so that we can log it here.

**Review by tjasko — CHANGES_REQUESTED** [review]

File: internal/server/kong/ws/config/configuration.go
Comment (omegabytes) [reply internal/server/kong/ws/config/configuration.go:60]: 1. [This is tracks total](https://github.com/Kong/koko/pull/485/files#diff-4b5bf8ad94aaa24b01cab08505b8ac550fc4efe48eac16233df472395212c06dR481), which accounts for success and fail states as well
2. I take this to mean `t=.001s` vs `t=1ms`. I'm accustomed to emitting ms, is emitting decimal seconds a prometheus convention?

File: internal/server/kong/ws/server.go
Comment (omegabytes) [reply internal/server/kong/ws/server.go:90]: @mikefero wdyt? You mentioned we should include request validation when we talked about this metric. I don't have a strong opinion either way.

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [reply internal/server/kong/ws/config/configuration.go:60]: 1. Yes, that's fine.
2. Derp, I forgot to link the doc. Yes, it's a convention. https://prometheus.io/docs/practices/naming/

File: internal/server/kong/ws/server.go
Comment (tjasko) [reply internal/server/kong/ws/server.go:90]: The strange thing to me is the naming, as the metric is named to make it think it's the connection duration only. It's fine for including the request validation, but it seemed a little strange.

Either way, both are part of establishing a connection. Whenever we switch this over to the Prometheus metric format, we can at least provide a help message documenting what it's capturing.

Comment (mikefero) [reply internal/server/kong/ws/server.go:90]: I am fine with having the name indicate this is a validated authenticated connection duration; more verbosity in the name doesn't hurt as we could add the other metrics during the connection phases later on.

File: internal/server/kong/ws/server.go
Comment (omegabytes) [reply internal/server/kong/ws/server.go:126]: I've seen us use a "protocol": "ws" metric tag in other places. I don't have an opinion either way.

File: internal/server/kong/ws/node.go
Comment (omegabytes) [reply internal/server/kong/ws/node.go:191]: I will address this as a separate issue.

**Comment by omegabytes** [comment]
> Should we use a `cp_` prefix for these metrics?

Are you suggesting something like `cp_data_plane_connect_duration_seconds`?

**Comment by GGabriele** [comment]
> > Should we use a `cp_` prefix for these metrics?
> 
> Are you suggesting something like `cp_data_plane_connect_duration_seconds`?

Yes, basically indicating which koko component is generating these metrics.

File: internal/server/kong/ws/server.go
Comment (tjasko) [reply internal/server/kong/ws/server.go]: > I've seen us use a "protocol": "ws" metric tag in other places. I don't have an opinion either way.
At least right now, there's only one protocol here, as it's all websockets anyway (for both wRPC & the older WS protocol). We don't need to rename it to be a bit more specific, was mostly just a question whether we should.

File: internal/server/kong/ws/server.go
Comment (tjasko) [reply internal/server/kong/ws/server.go:90]: Ohh I'm fine with how it is right now, just that it's slightly confusing as mentioned previously. Technically speaking this is the time for authentication & the actual connection. Perhaps instead of saying "connect", "authenticate" may be a better word.

File: internal/server/kong/ws/node.go
Comment (tjasko) [reply internal/server/kong/ws/node.go]: That works with me, thanks for removing this metric then.

File: internal/server/kong/ws/payload.go
Comment (tjasko) [inline internal/server/kong/ws/payload.go:144]: Did we want to update the status here to reflect whether `p.vc.ProcessConfigTableUpdates()` errored?

Additionally, this is capturing the duration of the config loading + storing as well, so it's technically more than just version compatibility, though at least right now, the majority of this code is focusing on versions compatibility right now. Perhaps well want to move this metric emitting code below the error handling logic for `p.vc.ProcessConfigTableUpdates()` & additionally move the `configUpdateProcessingStartTime` assignment right before the call. This is bit of a nit, as it shouldn't skew the metric results too badly,  but it could since this part of the code is blocked by a mutex.

Comment (omegabytes) [reply internal/server/kong/ws/payload.go]: Excluded caching from processing metric. We can look at instrumenting cache hit/miss/set/error behavior in the future.
[bb17fd1](https://github.com/Kong/koko/pull/485/commits/bb17fd17013470fbc21afa2117546948ca2e9d0d)

Comment (tjasko) [reply internal/server/kong/ws/payload.go]: Awesome, looks good with me! Thanks for moving the logic around, I was mostly concerned about the mutex throwing off the validity of this metric, so appreciate that change!

**Review by tjasko — DISMISSED** [review]

File: internal/server/kong/ws/config/configuration.go
Comment (hbagdi) [inline internal/server/kong/ws/config/configuration.go:6]: General comment: Are we sure we want to track at a seconds granularity? Isn't that too large?

File: internal/server/kong/ws/server.go
Comment (hbagdi) [inline internal/server/kong/ws/server.go:126]:
```suggestion
	metrics.Histogram("wrpc_connection_setup_duration_seconds",
```

Once again, do not prefix with DP or CP, those qualifiers are not necessary.

File: internal/server/kong/ws/config/configuration.go
Comment (omegabytes) [reply internal/server/kong/ws/config/configuration.go:62]: Histogram of `data_plane_config_mutation_individual_duration_seconds` represents count/sum/quantiles etc of all observations for this individual mutations. In my opinion, success and fail cases are a subtype of the metric and should be handled with tags. This is the same reason I use a mutator tag with the mutator name instead of appending the mutator name to the metric name.

To extend the reasoning a bit: imagine we have `config_mutation_individual_duration_seconds_fail` and want to emit metrics about why the failure happened. In this case, I'd include a `reason` tag with potential (example) values such as "internal" , "bad_input", "invalid_type", "timeout" etc. Or, I could append that reason to the metric and end up with 4 metric names for one very similar data point.

Extended further, if we append a reason to all our metrics then we are creating a ton of metric names. We also need to choose if we want to instrument those separate metrics in many places, or interpolate reason into the metric name to save on duplication of non-feature code.

File: internal/server/kong/ws/manager.go
Comment (omegabytes) [reply internal/server/kong/ws/manager.go:484]: This is an aggregate of https://github.com/Kong/koko/pull/485/files#diff-e0a3738e9a2e7a46592b82fb6819a25aa35998cefcec4dfe2e22a9b99cd0f48fR60, including `ReconfigurePayload` execution time. If we do not want to include `ReconfigurePayload` execution time, we can remove this and instead take the sum of `data_plane_config_mutation_individual_duration_seconds`

File: internal/server/kong/ws/config/configuration.go
Comment (omegabytes) [reply internal/server/kong/ws/config/configuration.go:6]: I agree - I am accustomed to ms as the base unit. However, @tjasko pointed out that Prometheus convention is to use seconds as the base unit: https://github.com/Kong/koko/pull/485#discussion_r1003676099

**Comment by omegabytes** [comment]
@hbagdi I've made the requested changes. Note that [I am following established convention in the code](https://github.com/Kong/koko/blob/4809bbf758d1bb92e4ed5290310d0b65e1341a43/internal/server/kong/ws/config_service.go#L67-L75) where possible, and review comments indicate we are no longer following that pattern.

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [reply internal/server/kong/ws/config/configuration.go:62]: The Prometheus [convention](https://prometheus.io/docs/practices/naming/) states:
> ...should represent the same logical thing-being-measured across all label dimensions.

Because we're logging config mutation duration here, regardless of if it passed/failed, we're still logging how long it took. Also because we're not conditionally adding any labels based on whether it passed/failed, this particular metric is okay IMHO. In the event we wanted to log more labels that was specific to only one edge-case (e.g.: when it fails), then I'd say it should be split out into its own metric.

This is honestly up for interpretation though, I've seen it done both ways, but that's my 2¢.

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [reply internal/server/kong/ws/config/configuration.go]: I did comment on something similar in a chat with Alex, and I do agree here, as the protocol doesn't impact this area of the code.

File: internal/server/kong/ws/config/configuration.go
Comment (tjasko) [reply internal/server/kong/ws/config/configuration.go]: We definitely do need to figure out a better standard for namespacing our metrics, as right now, it's not very well thought of. I've already brought this up with Alex regarding the metrics v2 implementation.

I could go either way here, the configuration is being pushed to the data plane, and keeping `data_plane` in it signals that it's the config being pushed to the DP, and not some other config. How can we make this metric name obvious in what it does when someone is reading all the metrics being emitted by Koko? How this metric is currently named helps with that.

File: internal/server/kong/ws/payload.go
Comment (tjasko) [reply internal/server/kong/ws/payload.go]: Given that we already have metrics called `version_compatibility_configuration_change_count`, perhaps we shouldn't deviate from the existing standards and just call it `version_compatibility_configuration_duration_seconds`.

As mentioned [previously](https://github.com/Kong/koko/pull/485#discussion_r1008457397), we do need to introduce some proper metric series namespacing, but we can get things cleaned up in the v2 integration & take it more seriously.

File: internal/server/kong/ws/server.go
Comment (tjasko) [reply internal/server/kong/ws/server.go]: This code isn't just limited to wRPC, and given that we have an existing metric named `data_plane_ping_total`, this is following current conventions, right?

Comment (mikefero) [reply internal/server/kong/ws/server.go]: It makes sense to keep these void of qualifiers. I know @omegabytes you have already committed the change to `wrpc_connection_setup_duration_seconds` but as @tjasko pointed out it does make sense to not include that qualifier either.

`connection_setup_duration_seconds` seems appropriate.

**Review by mikefero — APPROVED** [review]
One small comment I left around `wprc_connection_setup_duration_seconds`, but other than that this looks good to go.

@tjasko mentioned [this](https://github.com/Kong/koko/pull/485#discussion_r1008459160) but I think `configuration_transformation_duration_seconds` is acceptable and when v2 comes around we can address the `version_compatibility` namespacing.

Please get a +1 from @tjasko; all other items have been addressed.

**Review by tjasko — APPROVED** [review]

**Pull request merged**
